src/main/java/uk/ac/ebi/eva/commons/models/converters/data/DBObjectToVariantConverter.java

```diff
@@ -0,0 +1,250 @@
+        this.variantSourceEntryConverter = variantSourceEntryConverter;
+        this.variantAnnotationConverter = new DBObjectToVariantAnnotationConverter();
+        this.statsConverter = statsConverter;
+    }
+    
+    
+    @Override
+    public Variant convertToDataModelType(DBObject object) {
+        String chromosome = (String) object.get(CHROMOSOME_FIELD);
+        int start = (int) object.get(START_FIELD);
+        int end = (int) object.get(END_FIELD);
+        String reference = (String) object.get(REFERENCE_FIELD);
+        String alternate = (String) object.get(ALTERNATE_FIELD);
+        Variant variant = new Variant(chromosome, start, end, reference, alternate);
+        if (object.containsField(IDS_FIELD)) {
+            Object ids = object.get(IDS_FIELD);
+            variant.setIds(new HashSet<String>(((Collection<String>) ids)));
+        } else {
+            variant.setIds(new HashSet<String>());
+        }
+
+        // Transform HGVS: List of map entries -> Map of lists
+        BasicDBList mongoHgvs = (BasicDBList) object.get(HGVS_FIELD);
+        if (mongoHgvs != null) {
+            for (Object o : mongoHgvs) {
+                DBObject dbo = (DBObject) o;
+                variant.addHgvs((String) dbo.get(TYPE_FIELD), (String) dbo.get(NAME_FIELD));
+            }
+        }
+        
+        // Files
+        if (variantSourceEntryConverter != null) {
+            BasicDBList mongoFiles = (BasicDBList) object.get(FILES_FIELD);
+            if (mongoFiles != null) {
+                for (Object o : mongoFiles) {
+                    DBObject dbo = (DBObject) o;
+                    variant.addSourceEntry(variantSourceEntryConverter.convertToDataModelType(dbo));
+                }
+            }
+        }
+
+        // Annotations
+        DBObject mongoAnnotation = (DBObject) object.get(ANNOTATION_FIELD);
+        if (mongoAnnotation != null) {
+            VariantAnnotation annotation = variantAnnotationConverter.convertToDataModelType(mongoAnnotation);
+            annotation.setChromosome(variant.getChromosome());
+            annotation.setAlternativeAllele(variant.getAlternate());
+            annotation.setReferenceAllele(variant.getReference());
+            annotation.setStart(variant.getStart());
+            variant.setAnnotation(annotation);
+        }
+        
+        // Statistics
+        if (statsConverter != null && object.containsField(STATS_FIELD)) {
+            DBObject stats = (DBObject) object.get(STATS_FIELD);
+            statsConverter.convertCohortsToDataModelType(stats, variant);
+        }
+        return variant;
+    }
+
+    @Override
+    public DBObject convertToStorageType(Variant object) {
+        // Attributes easily calculated
+        BasicDBObject mongoVariant = new BasicDBObject("_id", buildStorageId(object))
+//                .append(IDS_FIELD, object.getIds())    //Do not include IDs.
```

Comment:
Improved comments. The idea is that it's not the same to build the update query and to convert the variant.

We are using `bulk.find(query).upsert().updateOne(update);` query to insert variants, and `update` is built as `update.append("$addToSet", addToSet).append("$setOnInsert", variantConverter.convert(variant));`.

The `addToSet` part includes the id, so it will be added if the variant document (and the ids array) already exist. If the variant is not found, the converted variant is inserted and then the addToSet part is performed.

We should probably refactor this part, as now we have a writer that makes half the query, a converter that makes half the query and half the conversion, and another converter that makes the rest of the conversion (coupled with the query design).